pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java

```diff
@@ -284,21 +284,35 @@ private void receiveMessageFromConsumer(ConsumerImpl<T> consumer) {
     // Must be called from the internalPinnedExecutor thread
     private void messageReceived(ConsumerImpl<T> consumer, Message<T> message) {
         checkArgument(message instanceof MessageImpl);
-        TopicMessageImpl<T> topicMessage = new TopicMessageImpl<>(consumer.getTopic(),
-                consumer.getTopicNameWithoutPartition(), message, consumer);
 
-        if (log.isDebugEnabled()) {
-            log.debug("[{}][{}] Received message from topics-consumer {}",
-                    topic, subscription, message.getMessageId());
-        }
+        // Mutually exclusive with redeliver to prevent the message
+        //from being cleared after being added to the incomingMessages
+        lock.readLock().lock();
```

Comment:
We can only add a check when polling out the message from the receiver queue, looks like the queue might have messages with lower epoch, but will be skipped when polling out from the queue.

I think it will simplify the implementation.